File: python/src/refactoring/golden_master.py
from dataclasses import dataclass
from typing import Optional, Literal
from datetime import datetime
import random


@dataclass
class OrderItem:
    sku: str
    description: str
    unit_price: float
    quantity: int
    category: Optional[Literal['general', 'food', 'books']] = None


@dataclass
class Order:
    id: str
    customer_name: str
    items: list[OrderItem]


class ReceiptPrinter:
    """Do not change this function at the beginning of the exercise; first create the Golden Master."""

    def print(self, order: Order) -> str:
        now = self.get_current_date()

        header = f"Recibo {order.id} - {now.strftime('%x')} {now.strftime('%X')}"

        subtotal = 0
        lines = []
        for idx, it in enumerate(order.items):
            line_total = round_money(it.unit_price * it.quantity)
            subtotal = round_money(subtotal + line_total)
            lines.append(f"{idx + 1}. {it.description} ({it.sku}) x{it.quantity} = ${line_total:.2f}")

        lucky_discount_pct = self.discount()
        lucky_discount = round_money(subtotal * lucky_discount_pct)

        taxable_general = sum(
            i.unit_price * i.quantity
            for i in order.items
            if i.category != 'books' and i.category != 'food'
        )
        food_tax = sum(
            i.unit_price * i.quantity * 0.03
            for i in order.items
            if i.category == 'food'
        )
        general_tax = taxable_general * 0.07
        taxes = round_money(general_tax + food_tax)

        total = round_money(subtotal - lucky_discount + taxes)

        summary = [
            f"Subtotal: ${subtotal:.2f}",
            (f"Descuento de la suerte: -${lucky_discount:.2f} ({lucky_discount_pct * 100:.2f}%)"
             if lucky_discount > 0
             else "Descuento de la suerte: $0.00 (0.00%)"),
            f"Impuestos: ${taxes:.2f}",
            f"TOTAL: ${total:.2f}",
        ]

        return '\n'.join([header, *lines, '---', *summary])

    def discount(self) -> float:
        lucky_discount_pct = 0
        if random.random() < 0.1:
            lucky_discount_pct = random.random() * 0.05
        return lucky_discount_pct

    def get_current_date(self) -> datetime:
        return datetime.now()


def round_money(n: float) -> float:
    return round(n * 100) / 100


# Products catalog
PRODUCTS: list[dict] = [
    {'sku': 'BK-001', 'description': 'Libro: Clean Code', 'unit_price': 30, 'category': 'books'},
    {'sku': 'FD-010', 'description': 'Café en grano 1kg', 'unit_price': 12.5, 'category': 'food'},
    {'sku': 'GN-777', 'description': 'Cuaderno A5', 'unit_price': 5.2, 'category': 'general'},
    {'sku': 'GN-123', 'description': 'Bolígrafos (pack 10)', 'unit_price': 3.9, 'category': 'general'},
    {'sku': 'FD-222', 'description': 'Té verde 200g', 'unit_price': 6.75, 'category': 'food'},
]

CUSTOMERS = ['Ana', 'Luis', 'Mar', 'Iván', 'Sofía']


# Utility to generate Orders
def generate_order(
    id: str,
    customer_name: str,
    num_items: int,
    quantity: int,
) -> Order:
    items = []
    for i in range(num_items):
        p = PRODUCTS[i]
        items.append(OrderItem(
            sku=p['sku'],
            description=p['description'],
            unit_price=p['unit_price'],
            quantity=quantity,
            category=p.get('category')
        ))

    return Order(id=id, customer_name=customer_name, items=items)

File: go/refactoring/golden_master.go
package refactoring

import (
	"fmt"
	"math/rand"
	"strings"
	"time"
)

// OrderItem represents an item in an order
type OrderItem struct {
	SKU         string
	Description string
	UnitPrice   float64
	Quantity    int
	Category    string // "general", "food", "books", or empty
}

// Order represents a customer order
type Order struct {
	ID           string
	CustomerName string
	Items        []OrderItem
}

// ReceiptPrinter prints receipts for orders
// Do not change this struct at the beginning of the exercise; first create the Golden Master.
type ReceiptPrinter struct {
	rng *rand.Rand
}

// NewReceiptPrinter creates a new ReceiptPrinter
func NewReceiptPrinter() *ReceiptPrinter {
	return &ReceiptPrinter{
		rng: rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// Print generates a receipt string for the given order
func (rp *ReceiptPrinter) Print(order Order) string {
	now := rp.GetCurrentDate()

	header := fmt.Sprintf("Recibo %s - %s", order.ID, now.Format("01/02/06 15:04:05"))

	subtotal := 0.0
	var lines []string
	for idx, it := range order.Items {
		lineTotal := roundMoney(it.UnitPrice * float64(it.Quantity))
		subtotal = roundMoney(subtotal + lineTotal)
		lines = append(lines, fmt.Sprintf("%d. %s (%s) x%d = $%.2f",
			idx+1, it.Description, it.SKU, it.Quantity, lineTotal))
	}

	luckyDiscountPct := rp.discount()
	luckyDiscount := roundMoney(subtotal * luckyDiscountPct)

	taxableGeneral := 0.0
	foodTax := 0.0
	for _, i := range order.Items {
		if i.Category != "books" && i.Category != "food" {
			taxableGeneral += i.UnitPrice * float64(i.Quantity)
		}
		if i.Category == "food" {
			foodTax += i.UnitPrice * float64(i.Quantity) * 0.03
		}
	}
	generalTax := taxableGeneral * 0.07
	taxes := roundMoney(generalTax + foodTax)

	total := roundMoney(subtotal - luckyDiscount + taxes)

	discountLine := ""
	if luckyDiscount > 0 {
		discountLine = fmt.Sprintf("Descuento de la suerte: -$%.2f (%.2f%%)",
			luckyDiscount, luckyDiscountPct*100)
	} else {
		discountLine = "Descuento de la suerte: $0.00 (0.00%)"
	}

	summary := []string{
		fmt.Sprintf("Subtotal: $%.2f", subtotal),
		discountLine,
		fmt.Sprintf("Impuestos: $%.2f", taxes),
		fmt.Sprintf("TOTAL: $%.2f", total),
	}

	var parts []string
	parts = append(parts, header)
	parts = append(parts, lines...)
	parts = append(parts, "---")
	parts = append(parts, summary...)

	return strings.Join(parts, "\n")
}

func (rp *ReceiptPrinter) discount() float64 {
	luckyDiscountPct := 0.0
	if rp.rng.Float64() < 0.1 {
		luckyDiscountPct = rp.rng.Float64() * 0.05
	}
	return luckyDiscountPct
}

// GetCurrentDate returns the current date/time
func (rp *ReceiptPrinter) GetCurrentDate() time.Time {
	return time.Now()
}

func roundMoney(n float64) float64 {
	return float64(int(n*100+0.5)) / 100
}

// Products catalog
var Products = []map[string]interface{}{
	{"sku": "BK-001", "description": "Libro: Clean Code", "unit_price": 30.0, "category": "books"},
	{"sku": "FD-010", "description": "Café en grano 1kg", "unit_price": 12.5, "category": "food"},
	{"sku": "GN-777", "description": "Cuaderno A5", "unit_price": 5.2, "category": "general"},
	{"sku": "GN-123", "description": "Bolígrafos (pack 10)", "unit_price": 3.9, "category": "general"},
	{"sku": "FD-222", "description": "Té verde 200g", "unit_price": 6.75, "category": "food"},
}

// Customers list
var Customers = []string{"Ana", "Luis", "Mar", "Iván", "Sofía"}

// GenerateOrder creates a test order
func GenerateOrder(id, customerName string, numItems, quantity int) Order {
	var items []OrderItem
	for i := 0; i < numItems && i < len(Products); i++ {
		p := Products[i]
		item := OrderItem{
			SKU:         p["sku"].(string),
			Description: p["description"].(string),
			UnitPrice:   p["unit_price"].(float64),
			Quantity:    quantity,
			Category:    p["category"].(string),
		}
		items = append(items, item)
	}

	return Order{
		ID:           id,
		CustomerName: customerName,
		Items:        items,
	}
}
